### ISSUE TITLE
Duplicate when importing

### ISSUE THREAD

**Issue by phzbox** [postcss/postcss-import#224]
Hi, I'm not sure if it's a bug or something I misunderstood.

With webpack, I import the same file in two different css files but it gets inlined twice.

```
A.css:
@import "core/cssutils.css";

B.css:
@import "core/cssutils.css";

core/cssutils.css:
body {
  font-family: verdana;
}

Webpack:
  postcss: function() {
    return [
      postcssImport
    ]
  },

Final result:
body{font-family:verdana}body{font-family:verdana}
```

**Comment by tonyxiao** [comment]
I'm seeing the same issue. @phzbox did you figure out a workaround?

**Comment by RyanZim** [comment]
@phzbox @tonyxiao Try the latest version, I suspect this issue is fixed.

**Comment by RyanZim** [comment]
Going to close this in a few days if I don't get a response.

**Comment by tomsapps** [comment]
bump for same issue

**Comment by RyanZim** [comment]
@tomsapps This issue should be fixed in v8.1.3 and later.

**Comment by bmenant** [comment]
@RyanZim I’m using v8.2.0 and I can reproduce the issue. Console log messages are also duplicated.

Or... could it be an erroneous webpack configuration?

```css
/* a.css */
@import "base.css";

/* b.css */
@import "base.css";

/* base.css */

:root {
  font-size: 1vmin;
}

h1 {
  /* pixrem module warns, since rem to pixel fallback with a dynamic root unit is not possible... */
  font-size: 1rem;
}
```

With this webpack config:

```js
    // Loaders...
    {
      test: /\.css$/,
      loader: extract('style-loader', [
        `css-loader?modules&localIdentName=${generateScopedName}&importLoaders=1`,
        'postcss-loader',
      ]),
    },
    // ...
  postcss: () => [
    atImport(),
    cssnext(),
  ],
```

The console outputs something like:

```    
Child extract-text-webpack-plugin:
    
    WARNING in ./~/css-loader?modules&localIdentName=[name]__[local]___[hash:base64:5]&importLoaders=1!./~/postcss-loader!./app/styles/base.css
    pixrem: Unit cannot be used for conversion, so 16px is used.

Child extract-text-webpack-plugin:
    
    WARNING in ./~/css-loader?modules&localIdentName=[name]__[local]___[hash:base64:5]&importLoaders=1!./~/postcss-loader!./app/styles/base.css
    pixrem: Unit cannot be used for conversion, so 16px is used.
    
```

And the extracted bundle.css:

```css
:root {
  font-size: 1vmin;
}
h1 { 
  /* pixrem module warns, since rem to pixel fallback with a dynamic root unit is not possible... */
  font-size: 16px;
  font-size: 1rem;
}
:root {
  font-size: 1vmin;
}
h1 {
  /* pixrem module warns, since rem to pixel fallback with a dynamic root unit is not possible... */
  font-size: 16px;
  font-size: 1rem;
}

```

**Comment by bmenant** [comment]
Note: using the Webpack [DedupePlugin](https://webpack.github.io/docs/list-of-plugins.html#dedupeplugin) does work out well for production bundles though.

**Comment by RyanZim** [comment]
@bmenant Since your imports are in separate files, I think this might be intended behavior. I'm not familiar with `postcss-loader`; @ai perhaps you can help here?

**Comment by RyanZim** [comment]
@ai Do you know what is the expected behavior here?

**Comment by ai** [comment]
Ouh. Yeap, logically `postcss-import` does everything right. Problem is because of different types of import. You import application CSS by `webpack`, but libraies by PostCSS.

But we need to give some sort of solution.

Maybe we could add some logic to `postcss-loader` to disable import if it is already imported in same chunk (but how we could know it). Maybe @mxstbr @MoOx or @michael-ciniawsky could give some good advice?

**Comment by MoOx** [comment]
Since when webpack does handle the import here?
postcss-import is executed before anything else so postcss-import should inline and then the others plugins do their job.

When using postcss-import , css-loader importLoaders=1 does not make sense (or I am drunk).

**Comment by MoOx** [comment]
The idea initially was to offer a dedup by default: for a single processing all identical files (eg: 2 files import "normalize.css") a single output should be emitted.

**Comment by ai** [comment]
Problem is when you import CSS from JS (by webpack), but in CSS import other CSS (by PostCSS)

**Comment by MoOx** [comment]
if you require( ) css files, obivoulsy you cannot expect a dedup by the process executed on the require(). It's probably a job for a smarter option and/or an optimisation plugin (remove duplicated parts in a big css)

**Comment by RyanZim** [comment]
> It's probably a job for a smarter option and/or an optimisation plugin (remove duplicated parts in a big css)

I'm with @MoOx on that one. As @bmenant said:

> Note: using the Webpack DedupePlugin does work out well for production bundles though.

TL;DR: If you're using webpack and `require` two different css files that both import the same css files, postcss-import won't dedupe it for you. Use https://webpack.github.io/docs/list-of-plugins.html#dedupeplugin for that.

Anyhow, gonna close this since this isn't the original issue topic anyhow.

**Comment by MoOx** [comment]
We have indeed [a test](https://github.com/postcss/postcss-import/blob/58ae15444661ec4d1c98f4168025a315e42cc5c4/test/import.js#L12-L14) to ensure there is [no duplicates](https://github.com/postcss/postcss-import/blob/58ae15444661ec4d1c98f4168025a315e42cc5c4/test/fixtures/no-duplicate.css) by [default](https://github.com/postcss/postcss-import/blob/58ae15444661ec4d1c98f4168025a315e42cc5c4/test/fixtures/no-duplicate.expected.css).

**Comment by bmenant** [comment]
@MoOx said:
> When using postcss-import , css-loader importLoaders=1 does not make sense (or I am drunk).

Well, I’m certainly doing it the wrong way then. Actually, I thought it might be the case. But it did not.
I guess because of the use the css `modules` option of css-loader. Without that `importLoaders=1` option, things weren’t working quite well with css-modules `composes from` statements (it’s like postcss didn’t process those imported css).

**Comment by pavel06081991** [comment]
I still have the same issue, here you can read about the problem https://github.com/postcss/postcss-loader/issues/63

It seems that there is no wat to fix it?

**Comment by michael-ciniawsky** [comment]
@pavel06081991 You can't use `postcss-import` in webpack for that, but this is specific to webpack. You add a 'module' when you load a `.css` file, which doesn't have any other context during that time as the file itself, `css-loader`s `importLoaders` transform the `@import` into a 'Module Request` and webpack can dedupe that later

`@import path/to/file` => `importLoaders` => `require('${...loaders}/path/to/file.css')`

if you use `postcss-import` for `@import` handling you will get bigger Blobs **per file**

**Comment by aaronbeall** [comment]
> @MoOx said:
> When using postcss-import , css-loader importLoaders=1 does not make sense (or I am drunk).

> @bmenant said:
> Well, I’m certainly doing it the wrong way then. Actually, I thought it might be the case. But it did not.
> I guess because of the use the css modules option of css-loader. Without that importLoaders=1 option, things weren’t working quite well with css-modules composes from statements (it’s like postcss didn’t process those imported css).

I have the same issue. When using CSS modules it seems I need *both* `postcss-import` and `importLoaders=1`... but trying to reconcile how these two work together is hard to wrap my mind around. Maybe `postcss-import` is pointless if you're already using `importLoaders`?

**Comment by michael-ciniawsky** [comment]
> Maybe postcss-import is pointless if you're already using importLoaders?

**Comment by michael-ciniawsky** [comment]
Don't use `postcss-import` especially when using CSS Modules (via `css-loader`)

`postcss-import` => `${import}${css}` (causes eventual duplicates (bloat), if the same `@import` is used across multiple file)

`css-loader?importLoaders=1` => 
```js
// Module Request, deduped by webpack 
import CSS__IMPORT__1 from 'path/to/import.css'

export default `
  ${css}
`
```

**Comment by aaronbeall** [comment]
@michael-ciniawsky When I remove `postcss-import` then a component CSS file that `@import`s from `colors.css` (to use CSS custom properties like `var(--mycolor)`) fails to work.

**Comment by michael-ciniawsky** [comment]
@aaronbeall use `postcss-custom-properties` instead

**Comment by aaronbeall** [comment]
@michael-ciniawsky I'm already using `postcss-cssnext` which, as I understand it, includes `postcss-custom-properties`? Still, even with `importLoaders=1` before my `postcss-loader` the vars only work if I also use `postcss-import`

**Comment by michael-ciniawsky** [comment]
Do you pass them via options as JSON to enhance the AST ?

```js
require('postcss-cssnext')({
  features: {
    customProperties: {
     variables: {
        color: "red",
        ...
      }
    }
  }
})

**Comment by aaronbeall** [comment]
No... I'm just using the `postcss-loader` and `postcss.config.js`. My setup is basically:

**webpack.config.js**
```
      { test: /\.css$/,
        exclude: /global\.css$/,
        use: [
          { loader: "style-loader", },
          { loader: "css-loader",
            options: {
              modules: true,
              camelCase: true,
              localIdentName: "[name]__[local]",
              sourceMap: process.env.NODE_ENV === "development",
              importLoaders: 1 
            }
          },
          { loader: "postcss-loader" } 
        ]
      }
```

**postcss.config.js**
```
module.exports = {
  plugins: {
    "postcss-import": {},
    "postcss-cssnext": {}
  }
}
```
If I remove `postcss-import` then in a CSS file I try to `@import` some custom properties like this:
```
@import "../colors.css";
.foo { color: var(--bg); }
```
I get a warning like:
> WARNING in ./node_modules/css-loader?{"modules":true,"camelCase":true,"localIdentName":"[name]__[local]","sourceMap":true,"importLoaders":1}!./node_modules/postcss-loader/lib!./src/components/App.css
(Emitted value instead of an instance of Error) postcss-custom-properties: /Users/abeall/Documents/github/aaronbeall/awesome-boilerplate/src/components/App.css:4:5: variable '--bg' is undefined and used without a fallback
 @ ./src/components/App.css 4:14-132
 @ ./src/components/App.tsx
 @ ./src/index.tsx
 @ multi (webpack)-dev-server/client?http://localhost:8080 ./src/index.tsx

**Comment by michael-ciniawsky** [comment]
**postcss.config.js**
```js
// colors.css => colors.json
const variables = require('path/to/colors.json')
// or 'inline'
// const variables = { ...variables }

module.exports = {
  plugins: {
    "postcss-import": {},
    "postcss-cssnext": {
        features: {
           customProperties: { variables }
        }
     }
  }
}

**Comment by aaronbeall** [comment]
Did you mean to leave `"postcss-import"` in that config? Seems you are saying that I convert `colors.css` to `colors.json` and provide it through the `postcss-cssnext` config?

**Comment by RyanZim** [comment]
Locking this old thread to prevent further noise. Pick up the discussion on https://gitter.im/postcss/postcss or in a new issue if this needs further discussion.
